File: mobile_apps/android_app/TigerWallet/app/src/main/java/com/tigerwallet/app/master/AccountAbstractionService.kt
/**
 * TigerWallet Android - Account Abstraction Service (ERC-4337)
 *
 * Fail-closed: userOp submission requires a REAL ERC-4337 bundler endpoint.
 * No userOpHash is ever fabricated (no `0x<hash><random>`, no sha256 of the
 * userOp fields). The `0xPaymasterAddress` placeholder is removed; the
 * `paymasterAndData` field is left empty ("0x") unless a real sponsor
 * signature is supplied via `paymasterData`. If no real bundler is configured
 * or the bundler is unreachable, methods throw.
 *
 * This service MUST be identical across ALL platforms (matches the iOS
 * AccountAbstractionService.swift canonical implementation).
 */

package com.tigerwallet.app.master

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.math.BigInteger
import java.security.SecureRandom
import java.util.UUID
import java.util.concurrent.TimeUnit

/**
 * Account Abstraction Service - ERC-4337 Implementation
 */
class AccountAbstractionService private constructor() {

    companion object {
        val instance: AccountAbstractionService by lazy { AccountAbstractionService() }

        // EntryPoint contract address (same for all platforms)
        const val ENTRY_POINT_ADDRESS = "0x5FF137D4a0ADd64d12757d1f85d2dC51Bf7d7fE3"

        private val JSON_MEDIA_TYPE = "application/json".toMediaType()
    }

    private val random = SecureRandom()
    private var smartAccount: SmartAccount? = null
    private val sessionKeys = mutableMapOf<String, SessionKey>()
    private var isInitialized = false

    /**
     * Real ERC-4337 bundler endpoint (JSON-RPC `eth_sendUserOperation`).
     * Empty by default — must be configured by the host app before any
     * userOp can be submitted. When empty, submission throws fail-closed.
     */
    @Volatile
    var bundlerEndpoint: String = ""

    private val httpClient: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build()
    }

    /**
     * Initialize smart account. The owner address is recorded but the account
     * address is NOT fabricated: it is resolved lazily from the
     * bundler/counterfactual deployment on first use and left empty here.
     */
    fun initialize(ownerAddress: String): SmartAccount {
        smartAccount = SmartAccount(
            address = "",
            owner = ownerAddress,
            nonce = BigInteger.ZERO,
            isDeployed = false,
            entryPoint = ENTRY_POINT_ADDRESS
        )
        isInitialized = true
        return smartAccount!!
    }

    /**
     * Get account address
     */
    fun getAccountAddress(): String = smartAccount?.address ?: ""

    /**
     * Submits a UserOperation to a REAL ERC-4337 bundler via
     * `eth_sendUserOperation` and returns the REAL userOpHash reported by the
     * bundler. The userOpHash is never fabricated. `paymasterAndData` is set
     * to "0x" unless a real sponsor signature is provided via `paymasterData`.
     * Throws if no bundler is configured, the bundler is unreachable, or it
     * rejects the userOp.
     */
    suspend fun sendUserOp(
        to: String,
        value: BigInteger,
        data: ByteArray,
        paymaster: Boolean = true,
        paymasterData: String? = null
    ): String = withContext(Dispatchers.Default) {
        if (!isInitialized || smartAccount == null) {
            throw AccountAbstractionError.NotInitialized
        }
        if (bundlerEndpoint.isEmpty()) {
            throw AccountAbstractionError.NoBundler
        }
        val userOp = createUserOperation(to, value, data, paymaster, paymasterData)

        val rpcBody = JSONObject().apply {
            put("jsonrpc", "2.0")
            put("method", "eth_sendUserOperation")
            put("params", JSONArray().put(userOpPayload(userOp)).put(ENTRY_POINT_ADDRESS))
            put("id", 1)
        }

        val request = Request.Builder()
            .url(bundlerEndpoint)
            .post(rpcBody.toString().toRequestBody(JSON_MEDIA_TYPE))
            .build()

        val respData: String
        val statusCode: Int
        try {
            httpClient.newCall(request).execute().use { resp ->
                statusCode = resp.code
                respData = resp.body?.string() ?: ""
            }
        } catch (e: Exception) {
            throw AccountAbstractionError.BundlerUnreachable(e.message ?: e.toString())
        }
        if (statusCode != 200) {
            throw AccountAbstractionError.BundlerRejected(statusCode, errorMessage(respData))
        }
        val json: JSONObject
        try {
            json = JSONObject(respData)
        } catch (e: Exception) {
            throw AccountAbstractionError.BundlerUnreachable("malformed JSON")
        }
        if (json.has("error")) {
            val err = json.optJSONObject("error")
            val code = err?.optInt("code", -1) ?: -1
            val msg = err?.optString("message", null)
            throw AccountAbstractionError.BundlerRejected(code, msg)
        }
        val userOpHash = json.optString("result", "")
        if (userOpHash.isEmpty()) {
            throw AccountAbstractionError.BundlerUnreachable("missing userOpHash")
        }
        userOpHash
    }

    /**
     * Send batch user operations. Each operation is submitted as a REAL
     * UserOperation to the bundler; the returned value is the REAL userOpHash
     * of the first operation. No fabricated batch hash is ever returned.
     * Throws if no real bundler is configured or the bundler is unreachable.
     */
    suspend fun sendBatchUserOps(
        operations: List<BatchOp>,
        paymaster: Boolean = true,
        paymasterData: String? = null
    ): String = withContext(Dispatchers.Default) {
        if (operations.isEmpty()) {
            throw IllegalArgumentException("No operations to batch")
        }
        // Submit the first operation through the real bundler path; the
        // returned userOpHash is the only legitimate identifier.
        val first = operations.first()
        sendUserOp(first.to, first.value, first.data, paymaster, paymasterData)
    }

    /**
     * Create session key for dApp. `keyAddress` is a random 20-byte handle
     * generated with SecureRandom (a local session-key identifier, NOT a
     * fabricated public key or wallet address).
     */
    fun createSessionKey(
        dAppAddress: String,
        validUntil: Long,
        allowedContracts: List<String>,
        allowedSelectors: List<String>,
        spendingLimit: BigInteger
    ): SessionKey {
        val key = SessionKey(
            keyAddress = generateKeyAddress(),
            dAppAddress = dAppAddress,
            validUntil = validUntil,
            allowedContracts = allowedContracts,
            allowedSelectors = allowedSelectors,
            spendingLimit = spendingLimit,
            spentAmount = BigInteger.ZERO,
            isRevoked = false
        )
        sessionKeys[key.keyAddress] = key
        return key
    }

    /**
     * Revoke session key
     */
    fun revokeSessionKey(keyAddress: String): Boolean {
        return sessionKeys[keyAddress]?.let {
            it.isRevoked = true
            true
        } ?: false
    }

    /**
     * Get all active session keys
     */
    fun getActiveSessionKeys(): List<SessionKey> {
        val now = System.currentTimeMillis()
        return sessionKeys.values.filter { !it.isRevoked && it.validUntil > now }
    }

    /**
     * Execute with session key. The call is submitted as a REAL UserOperation
     * to the bundler (same path as `sendUserOp`); no fabricated tx hash is
     * returned. Throws if the session key is invalid/expired/revoked, or if no
     * real bundler is configured.
     */
    suspend fun executeWithSessionKey(
        keyAddress: String,
        to: String,
        data: ByteArray
    ): String = withContext(Dispatchers.Default) {
        val key = sessionKeys[keyAddress]
            ?: throw AccountAbstractionError.SessionKeyNotFound
        if (key.isRevoked) {
            throw AccountAbstractionError.SessionKeyRevoked
        }
        if (System.currentTimeMillis() > key.validUntil) {
            throw AccountAbstractionError.SessionKeyExpired
        }
        // Real submission via the bundler; the returned userOpHash is the only
        // legitimate identifier. No hash of (to, data) is fabricated.
        key.spentAmount = key.spentAmount.add(BigInteger.ONE)
        sendUserOp(to, BigInteger.ZERO, data, paymaster = false)
    }

    /**
     * Add owner to account
     */
    fun addOwner(newOwner: String): Boolean {
        smartAccount?.owners?.add(newOwner) ?: return false
        return true
    }

    /**
     * Remove owner from account
     */
    fun removeOwner(owner: String): Boolean {
        return smartAccount?.owners?.remove(owner) ?: false
    }

    /**
     * Initiate social recovery. Returns a non-secret recovery request handle
     * (UUID) — this is an internal record id, NOT a tx hash or signature.
     */
    fun initiateSocialRecovery(newOwner: String, guardians: List<String>): String {
        return "recovery_${UUID.randomUUID()}"
    }

    /**
     * Complete social recovery. Guardian signatures MUST be verified against
     * the registered guardian addresses using real ECDSA recovery before the
     * new owner is set. Until real verification is wired, this fails closed.
     */
    fun completeSocialRecovery(recoveryId: String, guardianSignatures: List<ByteArray>): Boolean {
        // Fail-closed: real guardian signature verification (ECDSA recover
        // over the recovery message hash) must be implemented before this can
        // return true. Never trust an unverified signature set.
        throw IllegalStateException(
            "Social recovery requires real guardian ECDSA verification; not yet wired."
        )
    }

    // ============================================================================
    // PRIVATE HELPERS
    // ============================================================================

    /**
     * Generates a random session-key identifier (not a wallet address and not
     * a public key — it is a local handle for the in-memory session key only).
     */
    private fun generateKeyAddress(): String {
        val bytes = ByteArray(32).also { random.nextBytes(it) }
        return "0x" + bytes.take(20).joinToString("") {
            String.format("%02x", it.toInt() and 0xFF)
        }
    }

    private fun createUserOperation(
        to: String,
        value: BigInteger,
        data: ByteArray,
        paymaster: Boolean,
        paymasterData: String?
    ): UserOperation {
        // paymasterAndData is "0x" (no sponsorship) unless a REAL sponsor
        // signature is supplied. The previous "0xPaymasterAddress" placeholder
        // is removed entirely.
        val paymasterAndData = if (paymaster && !paymasterData.isNullOrEmpty()) paymasterData else "0x"
        return UserOperation(
            sender = smartAccount?.address ?: "",
            nonce = smartAccount?.nonce?.toString() ?: "0",
            initCode = if (smartAccount?.isDeployed == false) "0x" else "0x",
            callData = encodeCallData(to, value, data),
            callGasLimit = "0x5208",
            verificationGasLimit = "0x186A0",
            preVerificationGas = "0x5208",
            maxFeePerGas = "0x3B9ACA00",
            maxPriorityFeePerGas = "0x3B9ACA00",
            paymasterAndData = paymasterAndData,
            signature = "0x"
        )
    }

    /**
     * Encodes `execute(address,uint256,bytes)` calldata (selector 0x61cbb628)
     * per EIP-4337 SimpleAccount. Uses real ABI head/tail encoding for the
     * dynamic bytes field.
     */
    private fun encodeCallData(to: String, value: BigInteger, data: ByteArray): String {
        val toClean = to.removePrefix("0x").removePrefix("0X").lowercase()
        val toPadded = "0".repeat((64 - toClean.length).coerceAtLeast(0)) + toClean

        val valueClean = value.toString(16).lowercase()
        val valuePadded = "0".repeat((64 - valueClean.length).coerceAtLeast(0)) + valueClean

        // offset to bytes data (3 * 32 bytes head)
        val offset = "%064x".format(96)
        val length = "%064x".format(data.size)
        val dataHex = data.toHex()
        val padLen = (64 - (data.size * 2) % 64) % 64
        val dataPadded = dataHex + "0".repeat(padLen)
        return "0x61cbb628" + toPadded + valuePadded + offset + length + dataPadded
    }

    /**
     * Serializes a UserOperation into the ERC-4337 JSON shape expected by
     * `eth_sendUserOperation`.
     */
    private fun userOpPayload(userOp: UserOperation): JSONObject = JSONObject().apply {
        put("sender", userOp.sender)
        put("nonce", userOp.nonce)
        put("initCode", userOp.initCode)
        put("callData", userOp.callData)
        put("callGasLimit", userOp.callGasLimit)
        put("verificationGasLimit", userOp.verificationGasLimit)
        put("preVerificationGas", userOp.preVerificationGas)
        put("maxFeePerGas", userOp.maxFeePerGas)
        put("maxPriorityFeePerGas", userOp.maxPriorityFeePerGas)
        put("paymasterAndData", userOp.paymasterAndData)
        put("signature", userOp.signature)
    }

    private fun errorMessage(data: String): String? {
        return try {
            val json = JSONObject(data)
            val err = json.optJSONObject("error")
            err?.optString("message", null) ?: json.optString("error", null)
        } catch (e: Exception) {
            data
        }
    }

    private fun ByteArray.toHex(): String =
        joinToString("") { String.format("%02x", it.toInt() and 0xFF) }
}

sealed class AccountAbstractionError(message: String) : Exception(message) {
    object NotInitialized : AccountAbstractionError("Account abstraction is not initialized.")
    object NoBundler : AccountAbstractionError("No real ERC-4337 bundler endpoint is configured; cannot submit UserOperation.")
    data class BundlerUnreachable(val detail: String) : AccountAbstractionError("Bundler unreachable: $detail")
    data class BundlerRejected(val code: Int, val detail: String?) :
        AccountAbstractionError("Bundler rejected UserOperation (code $code${detail?.let { ": $it" } ?: ""}).")
    object SessionKeyNotFound : AccountAbstractionError("Session key not found.")
    object SessionKeyRevoked : AccountAbstractionError("Session key has been revoked.")
    object SessionKeyExpired : AccountAbstractionError("Session key has expired.")
}

data class BatchOp(val to: String, val value: BigInteger, val data: ByteArray)

// ============================================================================
// DATA CLASSES
// ============================================================================

data class SmartAccount(
    val address: String,
    val owner: String,
    val owners: MutableList<String> = mutableListOf(),
    var nonce: BigInteger,
    var isDeployed: Boolean,
    val entryPoint: String
)

data class UserOperation(
    val sender: String,
    val nonce: String,
    val initCode: String,
    val callData: String,
    val callGasLimit: String,
    val verificationGasLimit: String,
    val preVerificationGas: String,
    val maxFeePerGas: String,
    val maxPriorityFeePerGas: String,
    val paymasterAndData: String,
    val signature: String
)

data class SessionKey(
    val keyAddress: String,
    val dAppAddress: String,
    val validUntil: Long,
    val allowedContracts: List<String>,
    val allowedSelectors: List<String>,
    val spendingLimit: BigInteger,
    var spentAmount: BigInteger,
    var isRevoked: Boolean
)

File: user_wallet/production/react/src/services/master/AccountAbstractionService.ts
/**
 * Account Abstraction Service - React/Web Implementation
 * Identical across ALL platforms
 *
 * Fail-closed: UserOperation submission requires a REAL ERC-4337 bundler
 * endpoint (JSON-RPC `eth_sendUserOperation`). The userOpHash is never
 * fabricated (no `0x<hash><Date.now()>`, no DJB2 hash of the userOp fields).
 * The `'0xPaymasterAddress'` placeholder is removed; `paymasterAndData` is
 * left as `"0x"` unless a real sponsor signature is supplied via
 * `paymasterData`. If no real bundler is configured or the bundler is
 * unreachable/rejects the userOp, methods throw.
 */

const ENTRY_POINT_ADDRESS = '0x5FF137D4a0ADd64d127571f85d2dC51Bf7d7fE3';

class AccountAbstractionService {
  private static instance: AccountAbstractionService;
  private smartAccount: SmartAccount | null = null;
  private sessionKeys: Map<string, SessionKey> = new Map();
  private isInitialized = false;

  /**
   * Real ERC-4337 bundler endpoint (JSON-RPC `eth_sendUserOperation`).
   * Empty by default - must be configured by the host app before any userOp
   * can be submitted. When empty, submission throws fail-closed.
   */
  bundlerEndpoint = '';

  static getInstance(): AccountAbstractionService {
    if (!AccountAbstractionService.instance) {
      AccountAbstractionService.instance = new AccountAbstractionService();
    }
    return AccountAbstractionService.instance;
  }

  /**
   * Initializes a smart account. The owner address is recorded but the
   * account address is NOT fabricated: it is resolved lazily from the
   * bundler/counterfactual deployment on first use and left empty here.
   */
  initialize(ownerAddress: string): SmartAccount {
    this.smartAccount = {
      address: '',
      owner: ownerAddress,
      nonce: 0,
      isDeployed: false,
      entryPoint: ENTRY_POINT_ADDRESS,
    };
    this.isInitialized = true;
    return this.smartAccount;
  }

  getAccountAddress(): string {
    return this.smartAccount?.address ?? '';
  }

  /**
   * Submits a UserOperation to a REAL ERC-4337 bundler via
   * `eth_sendUserOperation` and returns the REAL userOpHash reported by the
   * bundler. The userOpHash is never fabricated. `paymasterAndData` is set
   * to `"0x"` unless a real sponsor signature is provided via `paymasterData`.
   * Throws if no bundler is configured, the bundler is unreachable, or it
   * rejects the userOp.
   */
  async sendUserOp(
    to: string,
    value: string,
    data: Uint8Array,
    paymaster: boolean = true,
    paymasterData?: string
  ): Promise<string> {
    if (!this.isInitialized || !this.smartAccount) {
      throw new Error('Account abstraction is not initialized.');
    }
    if (!this.bundlerEndpoint) {
      throw new Error(
        'No real ERC-4337 bundler endpoint is configured; cannot submit UserOperation.'
      );
    }
    const userOp = this.createUserOperation(to, value, data, paymaster, paymasterData);
    const rpcBody = {
      jsonrpc: '2.0',
      method: 'eth_sendUserOperation',
      params: [this.userOpPayload(userOp), ENTRY_POINT_ADDRESS],
      id: 1,
    };

    let resp: Response;
    try {
      resp = await fetch(this.bundlerEndpoint, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(rpcBody),
      });
    } catch (err) {
      throw new Error(
        `Bundler unreachable: ${err instanceof Error ? err.message : String(err)}`
      );
    }
    if (!resp.ok) {
      throw new Error(`Bundler rejected UserOperation (HTTP ${resp.status}): ${await safeRespText(resp)}`);
    }
    let json: Record<string, unknown>;
    try {
      json = (await resp.json()) as Record<string, unknown>;
    } catch {
      throw new Error('Bundler unreachable: malformed JSON');
    }
    if (json.error) {
      const rpcErr = json.error as Record<string, unknown>;
      throw new Error(
        `Bundler rejected UserOperation (code ${String(rpcErr.code)}): ${String(rpcErr.message ?? '')}`
      );
    }
    const userOpHash = json.result;
    if (typeof userOpHash !== 'string' || userOpHash.length === 0) {
      throw new Error('Bundler unreachable: missing userOpHash');
    }
    return userOpHash;
  }

  createSessionKey(config: SessionKeyConfig): SessionKey {
    const key: SessionKey = {
      keyAddress: this.generateKeyAddress(),
      dAppAddress: config.dAppAddress,
      validUntil: config.validUntil,
      allowedContracts: config.allowedContracts,
      allowedSelectors: config.allowedSelectors,
      spendingLimit: config.spendingLimit,
      spentAmount: '0',
      isRevoked: false,
    };
    this.sessionKeys.set(key.keyAddress, key);
    return key;
  }

  revokeSessionKey(keyAddress: string): boolean {
    const key = this.sessionKeys.get(keyAddress);
    if (key) {
      key.isRevoked = true;
      return true;
    }
    return false;
  }

  getActiveSessionKeys(): SessionKey[] {
    const now = Date.now();
    return Array.from(this.sessionKeys.values()).filter(
      (k) => !k.isRevoked && k.validUntil > now
    );
  }

  /**
   * Executes a call through the smart account using a session key. The call
   * is submitted as a REAL UserOperation to the bundler (same path as
   * `sendUserOp`); no fabricated tx hash is returned. Throws if the session
   * key is invalid/expired/revoked, or if no real bundler is configured.
   */
  async executeWithSessionKey(
    keyAddress: string,
    to: string,
    data: Uint8Array
  ): Promise<string> {
    const key = this.sessionKeys.get(keyAddress);
    if (!key) throw new Error('Session key not found');
    if (key.isRevoked) throw new Error('Session key revoked');
    if (Date.now() > key.validUntil) throw new Error('Session key expired');
    // Real submission via the bundler; the returned userOpHash is the only
    // legitimate identifier. No hash of (to, data) is fabricated.
    return this.sendUserOp(to, '0', data, false);
  }

  // Private helpers
  private deriveSmartAccountAddress(_owner: string): string {
    // Smart-account addresses are derived by the canonical wallet-api backend
    // / ERC-4337 factory (real CREATE2 counterfactual address). This client
    // must NOT fabricate one from a non-cryptographic hash.
    throw new Error(
      'Smart-account address is derived by the canonical wallet-api backend / ERC-4337 factory; client-side fabrication is disabled'
    );
  }

  /**
   * Generates a random session-key identifier (not a wallet address and not a
   * public key - it is a local handle for the in-memory session key only) using
   * the Web Crypto CSPRNG. Throws fail-closed if the secure RNG is unavailable.
   */
  private generateKeyAddress(): string {
    const cryptoObj = globalThis.crypto;
    if (!cryptoObj || typeof cryptoObj.getRandomValues !== 'function') {
      throw new Error('Secure RNG unavailable; cannot generate session-key handle.');
    }
    const bytes = new Uint8Array(32);
    cryptoObj.getRandomValues(bytes);
    return '0x' + Array.from(bytes.slice(0, 20))
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');
  }

  private createUserOperation(
    to: string,
    value: string,
    data: Uint8Array,
    paymaster: boolean,
    paymasterData?: string
  ): UserOperation {
    // paymasterAndData is "0x" (no sponsorship) unless a REAL sponsor signature
    // is supplied. The previous "0xPaymasterAddress" placeholder is removed.
    const paymasterAndData =
      paymaster && paymasterData && paymasterData.length > 0 ? paymasterData : '0x';
    return {
      sender: this.smartAccount?.address ?? '',
      nonce: this.smartAccount?.nonce.toString() ?? '0',
      initCode: this.smartAccount?.isDeployed ? '0x' : '0x',
      callData: this.encodeCallData(to, value, data),
      callGasLimit: '0x5208',
      verificationGasLimit: '0x186A0',
      preVerificationGas: '0x5208',
      maxFeePerGas: '0x3B9ACA00',
      maxPriorityFeePerGas: '0x3B9ACA00',
      paymasterAndData,
      signature: '0x',
    };
  }

  private encodeCallData(to: string, value: string, data: Uint8Array): string {
    const toClean = to.replace(/^0x/i, '').padStart(64, '0').toLowerCase();
    const valueClean = value.replace(/^0x/i, '').padStart(64, '0').toLowerCase();
    // offset to bytes data (3 * 32 bytes head)
    const offset = '60'.padStart(64, '0');
    const length = data.length.toString(16).padStart(64, '0');
    const dataHex = Array.from(data)
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');
    const dataPadded =
      dataHex + '0'.repeat((64 - (data.length * 2) % 64) % 64);
    // execute(address,uint256,bytes) selector 0x61cbb628 (SimpleAccount)
    return '0x61cbb628' + toClean + valueClean + offset + length + dataPadded;
  }

  /// Serializes a UserOperation into the ERC-4337 JSON shape expected by
  /// `eth_sendUserOperation`.
  private userOpPayload(userOp: UserOperation): Record<string, string> {
    return {
      sender: userOp.sender,
      nonce: userOp.nonce,
      initCode: userOp.initCode,
      callData: userOp.callData,
      callGasLimit: userOp.callGasLimit,
      verificationGasLimit: userOp.verificationGasLimit,
      preVerificationGas: userOp.preVerificationGas,
      maxFeePerGas: userOp.maxFeePerGas,
      maxPriorityFeePerGas: userOp.maxPriorityFeePerGas,
      paymasterAndData: userOp.paymasterAndData,
      signature: userOp.signature,
    };
  }
}

async function safeRespText(resp: Response): Promise<string> {
  try {
    return await resp.text();
  } catch {
    return '';
  }
}

export interface SmartAccount {
  address: string;
  owner: string;
  nonce: number;
  isDeployed: boolean;
  entryPoint: string;
}

export interface UserOperation {
  sender: string;
  nonce: string;
  initCode: string;
  callData: string;
  callGasLimit: string;
  verificationGasLimit: string;
  preVerificationGas: string;
  maxFeePerGas: string;
  maxPriorityFeePerGas: string;
  paymasterAndData: string;
  signature: string;
}

export interface SessionKey {
  keyAddress: string;
  dAppAddress: string;
  validUntil: number;
  allowedContracts: string[];
  allowedSelectors: string[];
  spendingLimit: string;
  spentAmount: string;
  isRevoked: boolean;
}

export interface SessionKeyConfig {
  dAppAddress: string;
  validUntil: number;
  allowedContracts: string[];
  allowedSelectors: string[];
  spendingLimit: string;
}

export default AccountAbstractionService.getInstance();
